### PR TITLE
[GHF] Print better merge rule match reason

### PR DESCRIPTION
**Pull request by malfet**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #74887
* #74893
* #74892

When running `find_matching_merge_rule` raise an exception that:
 - Contains generic message if files changed do not match any rules
 - Mentions rule that matches majority of files, but calls out few
   non-matching ones
 - Mentions matching rules, but calls out that it lacks reviewers
 - Mentions rule, but calls out missing mandatory checks

### CODE REVIEW

File: .github/scripts/trymerge.py [hunk 1]
```diff
@@ -579,46 +578,58 @@ def read_merge_rules(repo: GitRepo) -> List[MergeRule]:
...
     """Returns merge rule matching to this pr or raises an exception"""
     changed_files = pr.get_changed_files()
     approved_by = set(pr.get_approved_by())
     rules = read_merge_rules(repo)
+    reject_reason = f"PR {pr.pr_num} does not match merge rules"
+    #  Used to determine best rejection reason
+    # Score 0 to 10K - how many files rule matched
+    # Score 10K - matched all files, but no overlapping approvers
+    # Score 20K - matched all files and approvers, but lacks mandatory checks
+    reject_reason_score = 0
```
Comment (janeyx99) [inline .github/scripts/trymerge.py:591]:
```suggestion
    num_matching_files = 0
```
reject_reason_score is a confusing name here

Comment (janeyx99) [reply .github/scripts/trymerge.py:591]: oh nvm i see

File: .github/scripts/trymerge.py [hunk 2]
```diff
@@ -579,46 +578,58 @@ def read_merge_rules(repo: GitRepo) -> List[MergeRule]:
...
         rule_name = rule.name
         rule_approvers_set = set(rule.approved_by)
         patterns_re = patterns_to_regex(rule.patterns)
         approvers_intersection = approved_by.intersection(rule_approvers_set)
+        non_matching_files = []
+        for fname in changed_files:
+            if not patterns_re.match(fname):
+                non_matching_files.append(fname)
+        if len(non_matching_files) > 0:
+            score = len(changed_files) - len(non_matching_files)
```
Comment (janeyx99) [inline .github/scripts/trymerge.py:602]:
```suggestion
            num_matching_files = len(changed_files) - len(non_matching_files)
```
score is a vague variable--this is a nit

**Comment by malfet** [comment]
Applied both suggestions

**Review by osalpekar — APPROVED** [review]
Looks great - thanks!

**Review by janeyx99 — APPROVED** [review]

**Pull request closed without merging**